File: ScreenGuardian-Mobile/lib/services/reminder_manager.dart
/// Reminder Manager - Unified eye rest + posture change reminders
///
/// Design:
///   - Eye rest timer: fires every 20 minutes (configurable)
///   - Posture change: fires every 40 minutes (= 2× eye rest interval)
///   - When posture fires, show a COMBINED dialog: change posture + look 20ft away
///   - No separate posture timer — it piggybacks on the eye rest cycle

import 'dart:async';
import 'local_store.dart';

enum ReminderType {
  eyeRest,            // 20 min: just eye rest
  eyeRestAndPosture,  // 40 min: posture change + eye rest combined
}

class ReminderEvent {
  final ReminderType type;
  final int countdownSeconds;
  final bool meetingMode;

  ReminderEvent({
    required this.type,
    required this.countdownSeconds,
    required this.meetingMode,
  });

  bool get isCombined => type == ReminderType.eyeRestAndPosture;
}

class ReminderManager {
  final LocalStore _store;

  Timer? _timer;
  bool _active = false;
  int _triggerCount = 0; // counts eye rest triggers, every 2nd = posture

  int _eyeRestIntervalMs;
  bool _meetingMode;
  bool _eyeRestEnabled;
  bool _postureEnabled;

  final _controller = StreamController<ReminderEvent>.broadcast();
  Stream<ReminderEvent> get reminderStream => _controller.stream;

  ReminderManager({required LocalStore store})
      : _store = store,
        _eyeRestIntervalMs = 20 * 60 * 1000, // default 20 min
        _meetingMode = store.config.meetingMode,
        _eyeRestEnabled = store.config.eyeRestEnabled,
        _postureEnabled = store.config.postureEnabled;

  /// Reload settings from store (call after settings change)
  void reloadSettings() {
    final config = _store.config;
    _meetingMode = config.meetingMode;
    _eyeRestEnabled = config.eyeRestEnabled;
    _postureEnabled = config.postureEnabled;
    resetTimer();
  }

  void startTimers() {
    if (_eyeRestEnabled) _startTimer();
  }

  void stopTimers() {
    _timer?.cancel();
    _timer = null;
    _active = false;
  }

  void resetTimer() {
    stopTimers();
    startTimers();
  }

  void _startTimer() {
    _timer?.cancel();
    _timer = Timer(Duration(milliseconds: _eyeRestIntervalMs), () {
      if (_active) return;
      _active = true;
      _triggerCount++;

      // Every 2nd trigger → combined posture + eye rest
      final isPostureCycle = _postureEnabled && (_triggerCount % 2 == 0);

      _controller.add(ReminderEvent(
        type: isPostureCycle ? ReminderType.eyeRestAndPosture : ReminderType.eyeRest,
        countdownSeconds: isPostureCycle ? 120 : 20, // posture: 2min, eye rest: 20s
        meetingMode: _meetingMode,
      ));
    });
  }

  void onDialogClosed() {
    _active = false;
    _startTimer();
  }

  // Legacy compatibility
  void onEyeRestDialogClosed() => onDialogClosed();
  void onPostureDialogClosed() => onDialogClosed();

  bool get isActive => _active;
  bool get isEyeRestActive => _active;
  bool get isPostureActive => false; // no separate posture state
  bool get hasActiveReminder => _active;

  int get eyeRestIntervalMs => _eyeRestIntervalMs;
  int get postureIntervalMs => _eyeRestIntervalMs * 2; // derived
  bool get meetingMode => _meetingMode;
  bool get eyeRestEnabled => _eyeRestEnabled;
  bool get postureEnabled => _postureEnabled;

  void dispose() {
    stopTimers();
    _controller.close();
  }
}

File: ScreenGuardian-Desktop/src/services/reminder-manager.ts
/**
 * Reminder Manager - Unified eye rest + posture change reminders
 *
 * Design:
 *   - Single timer fires every 20 minutes (eye rest interval)
 *   - Every 2nd trigger (= 40 min) also includes posture change
 *   - Combined reminder: change posture + look 20ft away for 20s
 */

import { EventEmitter } from 'events';
import { LocalStore } from './local-store';

export enum ReminderType {
  EyeRest = 'eyeRest',
  EyeRestAndPosture = 'eyeRestAndPosture',
}

export interface ReminderEvent {
  type: ReminderType;
  countdownSeconds: number;
  meetingMode: boolean;
  isCombined: boolean;
}

export class ReminderManager extends EventEmitter {
  private store: LocalStore;
  private timer: NodeJS.Timeout | null = null;
  private _active = false;
  private _triggerCount = 0;

  private eyeRestIntervalMs: number;
  private meetingMode: boolean;
  private eyeRestEnabled: boolean;
  private postureEnabled: boolean;

  constructor(store: LocalStore) {
    super();
    this.store = store;
    this.eyeRestIntervalMs = 20 * 60 * 1000; // 20 minutes
    this.meetingMode = store.config.meetingMode;
    this.eyeRestEnabled = store.config.eyeRestEnabled;
    this.postureEnabled = store.config.postureEnabled;
  }

  reloadSettings(): void {
    const config = this.store.config;
    this.meetingMode = config.meetingMode;
    this.eyeRestEnabled = config.eyeRestEnabled;
    this.postureEnabled = config.postureEnabled;
    this.resetTimer();
  }

  startTimers(): void {
    if (this.eyeRestEnabled) this.startTimer();
  }

  stopTimers(): void {
    if (this.timer) { clearTimeout(this.timer); this.timer = null; }
    this._active = false;
  }

  resetTimer(): void {
    this.stopTimers();
    this.startTimers();
  }

  private startTimer(): void {
    if (this.timer) clearTimeout(this.timer);
    this.timer = setTimeout(() => {
      if (this._active) return;
      this._active = true;
      this._triggerCount++;

      const isPostureCycle = this.postureEnabled && (this._triggerCount % 2 === 0);

      this.emit('reminder', {
        type: isPostureCycle ? ReminderType.EyeRestAndPosture : ReminderType.EyeRest,
        countdownSeconds: isPostureCycle ? 120 : 20,
        meetingMode: this.meetingMode,
        isCombined: isPostureCycle,
      } as ReminderEvent);
    }, this.eyeRestIntervalMs);
  }

  onDialogClosed(): void {
    this._active = false;
    this.startTimer();
  }

  // Legacy compat
  onEyeRestDialogClosed(): void { this.onDialogClosed(); }
  onPostureDialogClosed(): void { this.onDialogClosed(); }

  get isActive(): boolean { return this._active; }
  get isEyeRestActive(): boolean { return this._active; }
  get isPostureActive(): boolean { return false; }
  get hasActiveReminder(): boolean { return this._active; }

  dispose(): void {
    this.stopTimers();
    this.removeAllListeners();
  }
}
